Issue 355: Clean up lombok annotations in API response models

ravikamaraj opened the pull request:
@wslulciuc , can you review. thanks. I would review the code coverage report, and add test cases

wslulciuc commented:
Thanks for the quick turn around on the issue, @ravikamaraj! I'll have time later this week to give your changes a review.

ravikamaraj commented:
@wslulciuc , I have added test cases for Preconditions.java, and refactored code in JobsResponse.java / JobRunsResponse.java, to improve code coverage. The code coverage is still failing, and has dropped for JobsResponse.java after adding the lombok annotations (@EqualsAndHashCode, @ToString). I don't see an oppurtunity to improve  code coverage, as annotations are not getting  included in the code coverage. Can you guide me on how to proceed further with this PR

thanks !

ravikamaraj commented:
issue-355 branch is outdated, I would pull latest commit from  master and apply my changes

ravikamaraj commented:
@wslulciuc , changes are complete, can you kindly review, thanks!

wslulciuc commented:
@ravikamaraj: Sure thing! I'll review your changes today.

ravikamaraj commented:
@wslulciuc , thank you for the feedback. Sure, will update shortly

wslulciuc commented on src/main/java/marquez/api/models/JobResponse.java:
```diff
@@ -14,26 +14,43 @@
+    this.name = checkNotBlank(name);
+    this.createdAt = checkNotBlank(createdAt);
+    this.updatedAt = checkNotBlank(updatedAt);
+    this.inputDatasetUrns = (List<String>) checkNotEmpty(inputDatasetUrns);
```
Both `JobResponse.inputDatasetUrns` and `JobResponse.outputDatasetUrns` may be empty. That is, it’s possible for a job to have only input, or only output datasets defined (see: https://marquezproject.github.io/marquez/#data-model)

Let's remove the assertion `MorePreconditions.checkNotEmpty()`. We can also achieve the same check with [`Preconditions.checkArgument()`](https://guava.dev/releases/snapshot/api/docs/com/google/common/base/Preconditions.html#checkArgument-boolean-). For example,

```java
void print(List<String> lines) {
  Preconditions.checkArgument(!lines.isEmpty, "lines must not be empty")
  ...
}
```

ravikamaraj replied:
sure

ravikamaraj replied:
done

wslulciuc commented on src/main/java/marquez/api/models/DatasourcesResponse.java:
```diff
@@ -20,5 +20,5 @@
 
 @Value
 public final class DatasourcesResponse {
```
_Minor:_ You can also remove `final` from the class definition:

```java
public class DatasourcesResponse {
  ...
}
```

ravikamaraj replied:
sure

ravikamaraj replied:
done

wslulciuc commented on src/main/java/marquez/api/models/JobsResponse.java:
```diff
@@ -15,14 +15,10 @@
 public class JobsResponse {
-  @Getter private List<JobResponse> jobs;
+  @NonNull private List<JobResponse> jobs;
 }
```
💯

wslulciuc commented on src/main/java/marquez/api/models/NamespaceResponse.java:
```diff
@@ -14,16 +14,30 @@
+      @NonNull String createdAt,
+      @NonNull String owner,
+      @Nullable String description) {
+    this.name = name;
```
Let's also ensure the values passed in are not blank

ravikamaraj replied:
sure

ravikamaraj replied:
done

wslulciuc commented on src/main/java/marquez/api/models/NamespaceResponse.java:
```diff
@@ -14,16 +14,30 @@
+  }
+
+  public Optional<String> getDescription() {
+    return Optional.ofNullable(this.description);
```
_Minor:_ The usage of `this` is redundant here.

ravikamaraj replied:
good catch. will update. thanks

ravikamaraj replied:
done

wslulciuc commented on src/main/java/marquez/api/models/JobRunResponse.java:
```diff
@@ -15,22 +15,27 @@
+      final String nominalStartTime,
+      final String nominalEndTime,
+      final String runArgs,
+      final String runState) {
```
`JobRunResponse.runState` should be a required field

ravikamaraj replied:
change has been implemented per review comment

ravikamaraj commented:
> With API models that wrap a list (ex: `DatasetsResponse`) we'll still want to use [`@Value`](https://projectlombok.org/features/Value) as it's the immutable variant of [`@Data`](https://projectlombok.org/features/Data).
> 
> For example, **class** `DatasetsResponse` should be:
> 
> ```java
> @Value
> public class DatasetsResponse {
>   @NonNull List<DatasetResponse> datasets;
> }
> ```
> 
> Mind updating the following class to be defined as above?
> 
>     * DatasourcesResponse
> 
>     * JobsResponse
> 
>     * JobRunsResponse
> 
>     * NamespacesResponse

below test cases are failing, while reading the entity. The test cases are getting passed, when no arg constructor is added. Can you review and let me know how do you want to handle this. thanks!
NamespaceIntegrationTest::testListNamespaceWithEmptyResultSet
JobResourceTest::testGetAllJobsInNamespace

wslulciuc commented on src/main/java/marquez/api/models/JobRunResponse.java:
```diff
@@ -15,22 +15,27 @@
+    this.nominalEndTime = nominalEndTime;
+    this.runArgs = runArgs;
+    this.runState = runState;
+  }
```
For the optional fields `nominalStartTime`, `nominalEndTime`, `runArgs` we'll want to define our own getter methods as follows:

```java
public Optional<String> getNominalStartTime() {
  return Optional.ofNullable(nominalStartTime);
}
...
```

ravikamaraj replied:
sure, will add  getter methods. Also should these fields be marked as `@Nullable` in constructor arguments

wslulciuc replied:
Yes! Those fields should be marked as `@Nullable` 🥇

ravikamaraj replied:
change has been implemented per review comment

wslulciuc commented on src/main/java/marquez/api/models/NamespaceResponse.java:
```diff
@@ -14,16 +14,32 @@
-@AllArgsConstructor
+import java.util.Optional;
+import javax.annotation.Nullable;
+import lombok.*;
```
_Minor:_ This would fail our style guide, see https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

ravikamaraj replied:
sure, will make note of it

ravikamaraj replied:
will make updates to other classes as well, where applicable

ravikamaraj replied:
change has been implemented per review comment

wslulciuc commented on src/main/java/marquez/api/models/JobsResponse.java:
```diff
@@ -17,12 +17,12 @@
 @Data
-@NoArgsConstructor
 @AllArgsConstructor
+@NoArgsConstructor
```
We should only annotate this class with `@Value`

wslulciuc replied:
Unless you added the additional annotations to get the tests to pass?

ravikamaraj replied:
yes, test cases were failing, so I had added these annotations. I would review the failing test cases, so as to handle `@value` annotation

ravikamaraj replied:
@wslulciuc, updated the classes to use `@Value` annotation, and fixed failing test cases.

ravikamaraj commented on src/test/java/marquez/api/resources/NamespaceIntegrationTest.java:
```diff
@@ -82,17 +82,26 @@ public void testBadCreateNamespaceRequest() {
             .get();
     assertEquals(Response.Status.OK.getStatusCode(), res.getStatus());
 
-    NamespacesResponse responseBody = res.readEntity(NamespacesResponse.class);
```
de serialization fails while using `@Value` annotation, because of no setters

wslulciuc commented:
@ravikamaraj, all these changes look great! But, before we can merge, we'll need to address a few, fairly minor, things:

1. Ensure all fields are of a primitive type. For example, `runId` is a `UUID` in **class** `JobRunResponse`, but should be of type  _String_
2. Ensure all fields are immutable. That is, made `final`.  For example, **class** `JobRunResponse` should be defined as:
   ```java
   public final class JobRunResponse {
     @Getter private final String runId;
     private final String nominalStartTime;
     ...
   }
   ```
3. Resolve any merge conflicts

wslulciuc commented:
> Will spend some time this week-end, to review if the test cases can be updated to work when the above classes are annotated with @Value. Can you let me know if you have solved similar problem, thanks

Yeah, this is something we certainly need to address. I''' opened an issue outlining the problem and how we want to go about updating our integration tests.

ravikamaraj commented:
> @ravikamaraj, all these changes look great! But, before we can merge, we'll need to address a few, fairly minor, things:
> 
>     1. Ensure all fields are of a primitive type. For example, `runId` is a `UUID` in **class** `JobRunResponse`, but should be of type  _String_
> 
>     2. Ensure all fields are immutable. That is, made `final`.  For example, **class** `JobRunResponse` should be defined as:
>        ```java
>        public final class JobRunResponse {
>          @Getter private final String runId;
>          private final String nominalStartTime;
>          ...
>        }
>        ```
> 
>     3. Resolve any merge conflicts

@wslulciuc , sure

ravikamaraj commented:
> > @ravikamaraj, all these changes look great! But, before we can merge, we'll need to address a few, fairly minor, things:
> > ```
> > 1. Ensure all fields are of a primitive type. For example, `runId` is a `UUID` in **class** `JobRunResponse`, but should be of type  _String_
> > 
> > 2. Ensure all fields are immutable. That is, made `final`.  For example, **class** `JobRunResponse` should be defined as:
> >    ```java
> >    public final class JobRunResponse {
> >      @Getter private final String runId;
> >      private final String nominalStartTime;
> >      ...
> >    }
> >    ```
> > 
> > 3. Resolve any merge conflicts
> > ```
> 
> @wslulciuc , sure

The above changes are complete, can you review. thanks

I would check on the code coverage drop by tomorrow, and provide an update

ravikamaraj commented:
closes issue #355
